### ISSUE TITLE
Improve memory locality for higher performances and reduced memory working set

### ISSUE THREAD

**Issue by padenot** [w3c/webcodecs#212]
We can add a second parameter to `decode`/`encode` and the output callbacks (for audio, video and images), that would be respectively an output buffer or handle, and the input buffer or handle (and also mirror it for encode vs. decode), to allow authors to reuse memory and avoid **lots** of memory traffic. GCs are amazing, but it's always better to not GC, reuse allocations, and always use the same bit of the address space, especially in very demanding multimedia scenarios.

This is **really** compelling whenever the input (resp. output for decode) is regular memory, since it means the working set is going to be pretty tight (for example help with https://github.com/w3c/webtransport/issues/231), but is really nice even when some buffers are backed by (say) GPU memory (because you're not doing malloc/free pairs for input/output buffers for respectively decoding and encoding). I'm not sure if GPU-memory backed surfaces can work in the same way.

Essentially, the proposal is to "pass through" the input to the output so that it can be reused. This reduces allocator use, but more importantly greatly improves memory locality.

When **not** using this new parameter, the implementation allocates as usual. This means that authors don't have to try to guess the buffer size, it's allocated once and then can be resized. The buffer is released naturally because it isn't referenced, immediately. In short, this is not a breaking change, but allows developers that want to control their memory allocation to be able to do so (and I believe this is something that is important for quite a few domains).

When the provided buffer is too small, the browser can allocate a new buffer, or use more complex tricks (realloc, or maybe it finds that the array buffer backing memory storage was in fact big enough, etc.).

When the provided buffer is way too big, the implementation can decide to swap it for a buffer that has a more appropriate size.

**Comment by aboba** [comment]
@padenot This sounds interesting.  Having some WebIDL might help in walking through the implications for various use cases (and related APIs, particularly those based on WHATWG Streams).  

A use case that comes to mind is media capture of a track, followed by WebCodecs encode and then sending (e.g. via WebTransport, RTCDataChannel, etc.). 

There is work underway to support [zero-copy capture](https://bugs.chromium.org/p/chromium/issues/detail?id=1120900). My understanding is that this will (in concert with [MediaCapture Transform](https://w3c.github.io/mediacapture-transform/)) provide a `VideoFrame` in a GPU buffer, and can then be provided to WebCodecs for Encoding, yielding an `encodedVideoChunk` that can then be sent via a transport. 

My understanding is that the GPU buffers are allocated and written to within the Media Capture implementation, so that WebCodecs encode can only take a reference to the GPU buffer as input.  So I'm wondering how the new parameters you are suggesting would interface with Streams-based APIs such as MediaCapture Transform and WebTransport.

**Comment by chcunningham** [comment]
Agree, sounds interesting! I _think_ I follow, but some quick pseudo code would help me be sure. 

Re: VideoFrame, one thing to note is that Chrome does do internal pooling of video frame resources for both GPU and CPU backed frames. For instance, [using this link](https://source.chromium.org/search?q=frame_pool%20f:decoder.cc&sq=) you can see the pooling done for a number of the internal decoders. Similarly, [here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/webcodecs/video_frame.cc;drc=9a42087863a44db3dd15cbca562b6d3fbc0fda01;l=592) you can see the CPU frame pool being used for WC VideoFrame's created via the planes constructor. When all references to the underlying frames are released, they go back to pool. 

We do similar pooling with audio decoders and the [backing memory for AudioData](https://source.chromium.org/search?q=AudioBufferMemoryPool%20f:decoder&sq=) (confusingly called AudioBuffer in our internal code).

Re: capture, this too uses pooling internally. [See here](https://source.chromium.org/chromium/chromium/src/+/main:media/capture/mojom/video_capture.mojom) for an overview of the IPC API that dishes out buffers and to observers (render processes) which then release them at some later point.

**Comment by aboba** [comment]
@chcunningham Thanks for the summary.  The argument seems to be that in the absence of GC, the pool may become fragmented, so that explicit locality can be beneficial.  For simple use cases (e.g. capture  -> encode) it seems like this could be avoided by a good pooling implementation (e.g. a GPU buffer could be reused soon after being freed). But perhaps for more complex scenarios memory leaks are more likely and GC will need to clean up the resulting mess.

**Comment by chcunningham** [comment]
Sounds good. I look forward to exploring it more. 

Another thing to mention: For VideoFrame's coming out of a VideoDecoder or the MediaStreamTrackProcessor, waiting for GC will actually stall the decode/camera stream. Authors really must call close() ASAP (and we warn them if they wait for GC) when they're done with a frame.

**Comment by padenot** [comment]
In terms of IDL, it would look something like this. I agree that what Chrome does is sensible (but I'm always impressed a how modern allocator make pools less useful than they used to be for regular memory), but this is also about decoder input and encoder outputs. I'm a bit wary of imposing pooling to developers, though.

<details>

```diff
@@ -5,9 +5,9 @@
   readonly attribute CodecState state;
   readonly attribute long decodeQueueSize;
 
   undefined configure(AudioDecoderConfig config);
-  undefined decode(EncodedAudioChunk chunk);
+  undefined decode(EncodedAudioChunk chunk, optional ArrayBuffer output);
   Promise<undefined> flush();
   undefined reset();
   undefined close();
 
@@ -18,9 +18,9 @@
   required AudioFrameOutputCallback output;
   required WebCodecsErrorCallback error;
 };
 
-callback AudioFrameOutputCallback = undefined(AudioFrame output);
+callback AudioFrameOutputCallback = undefined(AudioFrame output, BufferSource input);
 
 
 [Exposed=(Window,DedicatedWorker)]
 interface VideoDecoder {
@@ -29,9 +29,9 @@
   readonly attribute CodecState state;
   readonly attribute long decodeQueueSize;
 
   undefined configure(VideoDecoderConfig config);
-  undefined decode(EncodedVideoChunk chunk);
+  undefined decode(EncodedVideoChunk chunk, optional VideoFrame output);
   Promise<undefined> flush();
   undefined reset();
   undefined close();
 
@@ -42,9 +42,9 @@
   required VideoFrameOutputCallback output;
   required WebCodecsErrorCallback error;
 };
 
-callback VideoFrameOutputCallback = undefined(VideoFrame output);
+callback VideoFrameOutputCallback = undefined(VideoFrame output, BufferSource input);
 
 
 [Exposed=(Window,DedicatedWorker)]
 interface AudioEncoder {
@@ -53,9 +53,9 @@
   readonly attribute CodecState state;
   readonly attribute long encodeQueueSize;
 
   undefined configure(AudioEncoderConfig config);
-  undefined encode(AudioFrame frame);
+  undefined encode(AudioFrame frame, optional ArrayBuffer output);
   Promise<undefined> flush();
   undefined reset();
   undefined close();
 
@@ -66,9 +66,9 @@
   required EncodedAudioChunkOutputCallback output;
   required WebCodecsErrorCallback error;
 };
 
-callback EncodedAudioChunkOutputCallback = undefined(EncodedAudioChunk output);
+callback EncodedAudioChunkOutputCallback = undefined(EncodedAudioChunk output, ArrayBuffer input);
 
 
 [Exposed=(Window,DedicatedWorker)]
 interface VideoEncoder {
@@ -77,9 +77,9 @@
   readonly attribute CodecState state;
   readonly attribute long encodeQueueSize;
 
   undefined configure(VideoEncoderConfig config);
-  undefined encode(VideoFrame frame, optional VideoEncoderEncodeOptions options = {});
+  undefined encode(VideoFrame frame, optional VideoEncoderEncodeOptions options = {}, optional BufferSource input);
   Promise<undefined> flush();
   undefined reset();
   undefined close();
 
@@ -90,9 +90,9 @@
   required EncodedVideoChunkOutputCallback output;
   required WebCodecsErrorCallback error;
 };
 
-callback EncodedVideoChunkOutputCallback = undefined(EncodedVideoChunk output, VideoDecoderConfig? output_config);
+callback EncodedVideoChunkOutputCallback = undefined(EncodedVideoChunk output, VideoDecoderConfig? output_config, optional VideoFrame input);
 
 
 dictionary AudioDecoderSupport {
   boolean supported;
```

</details>

**Comment by chcunningham** [comment]
Triage note: marking 'extension', as the proposed interface changes extend existing methods/callbacks with extra arguments (not breaking).

**Comment by sandersdan** [comment]
In regards to the IDL proposal, I do not think this would be straightforward to implement. Specifically, during `VideoFrame` `encode()`, the resources could be in use by a different system (eg. be in use for a `copyTo()` operation, or displayed in a `<video>` via `MediaStreamTrackGenerator`). We could not return the underlying memory in that case.

What we could reliably do is implement a release callback that is provided at `VideoFrame` construction time, which would be called after all uses are finished.

**Comment by padenot** [comment]
As you note, this doesn't always work as-is. This proposal however always works when detaching inputs and I made in response to my comment in https://github.com/w3c/webcodecs/issues/104#issuecomment-832589627, but we can think about extending it. Another related issue is https://github.com/w3c/webcodecs/issues/287.

The general idea was to allow authors to write efficient code, but if we find a way to make it extra useful in all case, I'm all for it.

As usual, we need to be careful about not exposing GC, but that seems doable here. As long as we can specify the time the callback is called, this is fine to do -- essentially we can't rely on GC to call a callback.

**Comment by sandersdan** [comment]
> This proposal however always works when detaching inputs

I don't think I understand what you mean by this. Detaching the WebCodecs wrapper doesn't tell us anything about other uses of the backing resource, for example in the case that a frame has been `clone()`d.

**Comment by sandersdan** [comment]
You can absolutely detach a `VideoFrame` where the ref count on the media resource is greater than one:

```
let frame2 = frame1.clone();
encoder.encode(frame1, {transfer: [frame1]});  // detach frame1
// frame1 is detached but Chrome cannot return its memory until frame2 is closed
```

This can also happen by an async operation, or for other internal operations (eg. held while being displayed in a `<video>` via MSTG).

```
frame.copyTo(buf);  // start async operation
encoder.encode(frame, {transfer: [frame]});  // detach frame
// frame is detached but Chrome cannot return its memory until the copy completes
```

`VideoFrame` is more like `ArrayBufferView` than `ArrayBuffer`. (More precisely it follows the model of `ImageBitmap` but exposes a lot more state and explicit cloning.)

**Comment by erights** [comment]
> VideoFrame is more like `ArrayBufferView` than `ArrayBuffer`. (More precisely it follows the model of `ImageBitmap` but exposes a lot more state and explicit cloning.)

Would proposed [Immutable ArrayBuffers](https://github.com/tc39/proposal-immutable-arraybuffer) enable you to avoid such cloning?

**Comment by erights** [comment]
https://github.com/w3c/webcodecs/issues/212#issuecomment-2560348504

> > VideoFrame is more like `ArrayBufferView` than `ArrayBuffer`. (More precisely it follows the model of `ImageBitmap` but exposes a lot more state and explicit cloning.)
>
> Would proposed [Immutable ArrayBuffers](https://github.com/tc39/proposal-immutable-arraybuffer) enable you to avoid such cloning?

Ping. I am about to ask for advancement to stage 2.7 at tc39, so any timely answer or opinions would be most useful. Thanks!

**Comment by dalecurtis** [comment]
Yes immutable buffers would help remove a few copies and possibly allow us to expose the raw data layout. We'd want to be careful if we did that since it's very easy to work with the raw layout incorrectly (strides vs width/heights, visible vs coded, etc).
